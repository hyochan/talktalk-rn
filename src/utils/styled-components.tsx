import * as styledComponents from "styled-components/native";
import ThemeInterface from "./theme";

const {
  default: styled,
  css,
  ThemeProvider
} = styledComponents as styledComponents.ReactNativeThemedStyledComponentsModule<typeof ThemeInterface>;

export { css, ThemeProvider };
export default styled;