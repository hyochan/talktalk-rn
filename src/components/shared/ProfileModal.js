import LinearGradient from 'react-native-linear-gradient';
import React, {
  Component,
  useState,
  useImperativeHandle,
  forwardRef,
} from 'react';
import {
  ActivityIndicator,
  TouchableOpacity,
  StyleSheet,
  Image,
  Text,
  View,
  ViewStyle,
  ImageStyle,
  TextStyle,
  ImageSourcePropType,
} from 'react-native';
import Modal from 'react-native-modalbox';

import Icon5 from 'react-native-vector-icons/FontAwesome5';
import styled from 'styled-components/native';

import { User } from '../../models/User';
import { IC_MASK } from '../../utils/Icons';
import { ratio, colors } from '../../utils/Styles';
import { getString } from '../../../STRINGS';

const StyledWrapper = styled.View`
  height: 320;
  margin-horizontal: 20;
  aligin-self: stretch;
  align-items: center;
  justify-content: space-between;
`;

const StyledView = styled.View`
  marginTop: 40;
`;

const StyledViewBtns = styled.View`
  height: 80;
  align-self: stretch;
  background-color: white;

  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

const StyledViewBtnDivider = styled.View`
  width: 1;
  height: 80;
  background-color: ${colors.paleGray};
`;

const StyledTextDisplayName = styled.Text`
  font-size: 24;
  color: white;
  font-weight: bold;
  margin-top: 32;
  align-self: center;
`;

const StyledTextStatusMsg = styled.Text`
  font-size: 12;
  color: white;
  margin-top: 8;
  align-self: center;
`;

const StyledTextBtn = styled.Text`
  color: ${colors.dodgerBlue};
  font-size: 16;
`;

const StyledTextFriendAdded = styled.Text`
  color: white;
  font-size: 12;
  backround-color: ${colors.dusk};
  padding: 4;
`;

const StyledTextFriendAlreadyAdded = styled.Text`
  color: red;
  font-size: 12;
  background-color: ${colors.cloudyBlue};
  padding: 4;
`;

type Props = {
  onChatPressed?: () => void;
};

type Ref = {
  open: () => void;
  close: () => void;
  addFriend: () => void;
  deleteFriend: () => void;
};

type Styles = {
  viewBtn: ViewStyle,
};

const styles = {
  wrapper: {
    backgroundColor: 'transparent',
    alignSelf: 'stretch',
    height: 320,
    width: '90%',

    alignContent: 'center',
    alignItems: 'center',
    justifyContent: 'center',
  },
  viewBtn: {
    width: '50%',
    height: 80,
    alignItems: 'center',
    justifyContent: 'center',
  },
};

function Shared(props: Props, ref: Ref) {
  let modal: any;
  const [showAddBtn, setShowAddBtn] = useState(true);
  const [isAdding, setIsAdding] = useState(false);
  const [isFriendAdded, setIsFriendAdded] = useState(false);
  const [isFriendAlreadyAdded, setIsFriendAlreadyAdded] = useState(false);
  const [user, setUser] = useState(new User());

  const open = () => {
    setIsFriendAdded(false);
    setIsFriendAlreadyAdded(false);
    if (modal) {
      modal.open();
    }
  };

  const close = () => {
    if (modal) {
      modal.close();
    }
  };

  const addFriend = () => {

  };

  const deleteFriend = () => {
    if (modal) {
      modal.close();
    }
  };

  useImperativeHandle(ref, () => ({
    open,
    close,
    setUser: (user: User) => {
      setUser(user);
    },
    showAddBtn: (flag: boolean) => {
      setShowAddBtn(flag);
    },
  }));

  return (
    <Modal
      ref={(v) => modal = v}
      backdropOpacity={0.075}
      entry={'top'}
      position={'center'}
      style={styles.wrapper}
    >
      <LinearGradient
        style={{
          height: 320,
          marginHorizontal: 20,
          alignSelf: 'stretch',

          alignItems: 'center',
          justifyContent: 'space-between',
        }}
        start={{ x: 0.4, y: 0.6 }} end={{ x: 1.0, y: 0.8 }}
        locations={[0, 0.85]}
        colors={[colors.dodgerBlue, 'rgb(100,199,255)']}
      >
        <StyledView>
          <TouchableOpacity
            activeOpacity={0.5}
            // onPress={goToProfile}
          >
            {
              user.photoURL
                ? <Image style={{
                  alignSelf: 'center',
                }} source={user.photoURL}/>
                : <Icon5 name="meh" size={80} color={colors.dusk} light/>
            }
          </TouchableOpacity>
          <StyledTextDisplayName>{user.displayName}</StyledTextDisplayName>
          <StyledTextStatusMsg>{user.statusMsg}</StyledTextStatusMsg>
        </StyledView>
        {
          isFriendAdded
            ? <StyledTextFriendAdded>{getString('FRIEND_ADDED')}</StyledTextFriendAdded>
            : isFriendAlreadyAdded
              ? <StyledTextFriendAlreadyAdded>{getString('FRIEND_ALREADY_ADDED')}</StyledTextFriendAlreadyAdded>
              : null
        }
        <StyledViewBtns>
          <TouchableOpacity
            testID='btn-add-or-delete'
            activeOpacity={0.5}
            onPress={showAddBtn ? addFriend : deleteFriend}
            style={styles.viewBtn}
          >
            <View style={styles.viewBtn}>
              <StyledTextBtn>{
                showAddBtn ? getString('ADD_FRIEND') : getString('DELETE_FRIEND')
              }</StyledTextBtn>
            </View>
          </TouchableOpacity>
          <StyledViewBtnDivider />
          <TouchableOpacity
            testID='btn-chat'
            activeOpacity={0.5}
            onPress={props.onChatPressed}
            style={styles.viewBtn}
          >
            <View style={styles.viewBtn}>
              <StyledTextBtn>{getString('CHAT')}</StyledTextBtn>
            </View>
          </TouchableOpacity>
        </StyledViewBtns>
      </LinearGradient>
    </Modal>
  );
}

export default forwardRef(Shared);
