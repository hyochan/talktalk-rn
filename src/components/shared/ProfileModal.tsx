import LinearGradient from 'react-native-linear-gradient';
import React, {
  Component,
  useState,
  useImperativeHandle,
  forwardRef,
  Ref,
  RefObject,
} from 'react';
import {
  ActivityIndicator,
  TouchableOpacity,
  StyleSheet,
  Image,
  Text,
  View,
  ViewStyle,
  ImageStyle,
  TextStyle,
  ImageSourcePropType,
} from 'react-native';
import Modal from 'react-native-modalbox';
import Icon5 from 'react-native-vector-icons/FontAwesome5';
import styled, { ThemeProps, DefaultTheme, withTheme } from 'styled-components/native';
import { User } from '../../models/User';
import { IC_MASK } from '../../utils/Icons';
import { getString } from '../../../STRINGS';

const StyledView = styled.View`
  margin-top: 40px;
`;

const StyledViewBtns = styled.View`
  height: 80px;
  align-self: stretch;
  background-color: white;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

const StyledViewBtnDivider = styled.View`
  width: 1px;
  height: 80px;
  background-color: ${({ theme: { colors: { paleGray }}}) => paleGray};
`;

const StyledTextDisplayName = styled.Text`
  font-size: 24px;
  color: white;
  font-weight: bold;
  margin-top: 32px;
  align-self: center;
`;

const StyledTextStatusMsg = styled.Text`
  font-size: 12px;
  color: white;
  margin-top: 8px;
  align-self: center;
`;

const StyledTextBtn = styled.Text`
  color: ${({ theme: { colors: { dodgerBlue }}}) => dodgerBlue};
  font-size: 16px;
`;

const StyledTextFriendAdded = styled.Text`
  color: white;
  font-size: 12px;
  background-color: ${({ theme: { colors: { dusk }}}) => dusk};
  padding: 4;px
`;

const StyledTextFriendAlreadyAdded = styled.Text`
  color: red;
  font-size: 12px;
  background-color: ${({ theme: { colors: { cloudyBlue }}}) => cloudyBlue};
  padding: 4px;
`;

interface IProps extends ThemeProps<DefaultTheme> {
  testID?: string;
  ref?: any;
  onChatPressed?: () => void;
}

interface IRef {
  open: () => void;
  close: () => void;
  setUser: (user: User) => void;
  showAddBtn: (show: boolean) => void;
}

interface IStyles {
  wrapper: ViewStyle;
  viewBtn: ViewStyle;
}

const styles: IStyles = {
  wrapper: {
    backgroundColor: 'transparent',
    alignSelf: 'stretch',
    height: 320,
    width: '90%',
    alignContent: 'center',
    alignItems: 'center',
    justifyContent: 'center',
  },
  viewBtn: {
    width: '50%',
    height: 80,
    alignItems: 'center',
    justifyContent: 'center',
  },
};

const Shared = forwardRef<IRef, IProps>((props, ref) => {
  let modal: any;
  const [showAddBtn, setShowAddBtn] = useState(true);
  const [isAdding, setIsAdding] = useState(false);
  const [isFriendAdded, setIsFriendAdded] = useState(false);
  const [isFriendAlreadyAdded, setIsFriendAlreadyAdded] = useState(false);
  const [user, setUser] = useState(new User());

  const open = () => {
    setIsFriendAdded(false);
    setIsFriendAlreadyAdded(false);
    if (modal) {
      modal.open();
    }
  };

  const close = () => {
    if (modal) {
      modal.close();
    }
  };

  const addFriend = () => {

  };

  const deleteFriend = () => {
    if (modal) {
      modal.close();
    }
  };

  useImperativeHandle(ref, () => ({
    open,
    close,
    setUser: (newUser: User) => {
      setUser(newUser);
    },
    showAddBtn: (flag: boolean) => {
      setShowAddBtn(flag);
    },
  }));
  const { 
    photoURL,
    displayName,
    statusMsg
  } = user;
  const {
    theme: {
      colors: {
        dusk,
        dodgerBlue
      }
    }
  } = props;
  const imageURL = typeof photoURL === 'string' ? { uri: photoURL } : photoURL;
  return (
    <Modal
      ref={(v: any) => modal = v}
      backdropOpacity={0.075}
      entry={'top'}
      position={'center'}
      style={styles.wrapper}
    >
      <LinearGradient
        style={{
          height: 320,
          marginHorizontal: 20,
          alignSelf: 'stretch',
          alignItems: 'center',
          justifyContent: 'space-between',
        }}
        start={{ x: 0.4, y: 0.6 }} end={{ x: 1.0, y: 0.8 }}
        locations={[0, 0.85]}
        colors={[dodgerBlue, 'rgb(100,199,255)']}
      >
        <StyledView>
          <TouchableOpacity
            activeOpacity={0.5}
            // onPress={goToProfile}
          >
            {photoURL
              ? <Image 
                  style={{ alignSelf: 'center' }} 
                  source={imageURL} 
                />
              : <Icon5 name='meh' size={80} color={dusk} light/>
            }
          </TouchableOpacity>
          <StyledTextDisplayName>{displayName}</StyledTextDisplayName>
          <StyledTextStatusMsg>{statusMsg}</StyledTextStatusMsg>
        </StyledView>
        {isFriendAdded
          ? <StyledTextFriendAdded>{getString('FRIEND_ADDED')}</StyledTextFriendAdded>
          : isFriendAlreadyAdded
            ? <StyledTextFriendAlreadyAdded>{getString('FRIEND_ALREADY_ADDED')}</StyledTextFriendAlreadyAdded>
            : null
        }
        <StyledViewBtns>
          <TouchableOpacity
            testID='btn-add-or-delete'
            activeOpacity={0.5}
            onPress={showAddBtn ? addFriend : deleteFriend}
            style={styles.viewBtn}
          >
            <View style={styles.viewBtn}>
              <StyledTextBtn>{
                showAddBtn ? getString('ADD_FRIEND') : getString('DELETE_FRIEND')
              }</StyledTextBtn>
            </View>
          </TouchableOpacity>
          <StyledViewBtnDivider />
          <TouchableOpacity
            testID='btn-chat'
            activeOpacity={0.5}
            onPress={props.onChatPressed}
            style={styles.viewBtn}
          >
            <View style={styles.viewBtn}>
              <StyledTextBtn>{getString('CHAT')}</StyledTextBtn>
            </View>
          </TouchableOpacity>
        </StyledViewBtns>
      </LinearGradient>
    </Modal>
  );
});

export default withTheme(Shared);
