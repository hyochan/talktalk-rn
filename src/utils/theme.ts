import { Dimensions, Platform } from 'react-native';
import { getStatusBarHeight } from 'react-native-status-bar-height';
import { DefaultTheme } from 'styled-components';

const { width, height } = Dimensions.get('window');
let calRatio = width <= height ? 16 * (width / height) : 16 * (height / width);
if (width <= height) {
  if (calRatio < 9) {
    calRatio = width / 9;
  } else {
    calRatio = height / 18;
  }
} else {
  if (calRatio < 9) {
    calRatio = height / 9;
  } else {
    calRatio = width / 18;
  }
}

const theme: DefaultTheme = {
  screenWidth: width,
  screenHeight: height,
  ratio: calRatio / (360 / 9),
  statusBarHeight: getStatusBarHeight(true),
  colors: {
    background: '#e3e3e3',
    dodgerBlue: 'rgb(58,139,255)',
    dusk: 'rgb(65,77,107)',
    blueyGray: 'rgb(134,154,183)',
    cloudyBlue: 'rgb(175,194,219)',
    greenishCyan: 'rgb(80,227,194)',
    paleGray: 'rgb(233,237,244)',
  }
};

export default theme;