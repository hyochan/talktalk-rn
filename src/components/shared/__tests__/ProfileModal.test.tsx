import 'react-native';
import * as React from 'react';
import ProfileModal from '../ProfileModal';
import { ThemeProvider } from 'styled-components/native';
import theme from '../../../utils/theme';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { render, fireEvent } from 'react-native-testing-library';

const component: React.ReactElement = (
  <ThemeProvider theme={theme}>
    <ProfileModal theme={theme} />
  </ThemeProvider>
);

describe('[ProfileModal] rendering test', () => {
  it('renders as expected', () => {
    const json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
  });
});

describe('[ProfileModal] interaction', () => {
  let rendered: renderer.ReactTestRenderer;
  let testingLib: any;

  beforeAll(() => {
    rendered = renderer.create(component);
    testingLib = render(component);
  });

  it('should call [addFriend] when [btn-add-or-delete] has pressed', () => {
    // const addOrDeleteBtn = testingLib.getByTestId('btn-add-or-delete');
    // fireEvent(addOrDeleteBtn, 'press');
  });

  it('should call [onChatPressed] when [onChatPressed] has pressed', () => {
    // const chatBtn = testingLib.getByTestId('btn-chat');
    // fireEvent(chatBtn, 'press');
  });
});
