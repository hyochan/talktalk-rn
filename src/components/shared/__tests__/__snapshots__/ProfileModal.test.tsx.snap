// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ProfileModal] rendering test renders as expected 1`] = `<View />`;
